akka-http2-support/src/main/scala/akka/http/impl/http2/FrameParser.scala

```diff
@@ -0,0 +1,113 @@
+package akka.http.impl.http2
+
+import akka.http.impl.http2.Protocol._
+import FrameType._
+import akka.stream.Attributes
+import akka.stream.impl.io.ByteStringParser
+import akka.stream.stage.GraphStageLogic
+
+object FrameParser extends ByteStringParser[FrameEvent] {
+  import ByteStringParser._
+
+  abstract class Step extends ParseStep[FrameEvent]
+
+  def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new ParsingLogic {
+      startWith(ReadPreface)
+
+      object ReadPreface extends Step {
+        def parse(reader: ByteReader): ParseResult[FrameEvent] =
+          if (reader.remainingSize < 24) throw NeedMoreData
+          else if (reader.take(24) == Protocol.ConnectionPreface)
+            ParseResult(None, ReadFrame, false)
+          else
+            throw new RuntimeException("Expected ConnectionPreface!")
+
+      }
+
+      object ReadFrame extends Step {
+        def parse(reader: ByteReader): ParseResult[FrameEvent] = {
+          val length = reader.readShortBE() << 8 | reader.readByte()
+          val tpe = reader.readByte() // TODO: make sure it's valid
+          val flags = reader.readByte()
+          val streamId = reader.readIntBE()
+          // TODO: assert that reserved bit is 0 by checking if streamId > 0
+          val payload = reader.take(length)
+          val frame = parseFrame(FrameType.byId(tpe), flags, streamId, new ByteReader(payload))
+
+          ParseResult(Some(frame), ReadFrame, false)
+        }
+      }
+    }
+
+  def parseFrame(tpe: FrameType, flags: Int, streamId: Int, payload: ByteReader): FrameEvent = {
+    def isSet(flag: Int): Boolean = (flags & flag) != 0
+
+    tpe match {
```

Comment:
Good, no need to address right away, sprinkling around TODOs is a good idea during the PoC anyway